### ISSUE TITLE
Schema questions

### ISSUE THREAD

**Issue by gmaclennan** [digidem/mapeo-schema#36]
### Protobuf

- Store `links` as `Array<{ coreKey: Buffer, index: number }>`?
- Store `created_at` and `timestamp` as Timestamp or number vs. string?
- Rename `timestamp` to `edited_at`?
- Camelcase props? e.g. `createdAt`?
- `deviceId` as bytes
- Define type of observation refs
- Define type of observation attachments (since we can store metadata in the blobstore now, how much info here?)

### JSON Schema

- `created_at` and `timestamp` as type `number` or `Date`?
- Do we need to return the `schemaVersion` field? Maybe the contract with the client is that the latest schema version is always returned, and is typed.
- Also store `version` for `refs`?
- Restrictions on `tags`? E.g. JSON? Nested?

### Other types

- Protobuf for `node`, `way`, `relation`, `preset`, `field`?
- `device` and `project` record types needed?
- Anything else? `ownership`, `capability`?
- Mapping older schema versions - can do most at migration to avoid runtime migration?

**Comment by sethvincent** [comment]
> Store links as Array<{ coreKey: Buffer, index: number }>

Not needing to split on `@` sounds awesome to me.

> Define type of observation attachments

It seems like this could now just be whatever we need to do a get. Or whatever we need to fetch the metadata.

> Restrictions on tags? E.g. JSON? Nested?

Is there config with json value tags in use now? Nested json is kinda scary.

---

Yeah timestamp should be renamed to either updated_at or edited_at. updated_at is more common.

Using number type for any date fields makes sense to me.

Lately I keep daydreaming about how nice it would be if everything in js was snake_case. It's so readable! But, for consistency I'd lean toward making things camelCase.

**Comment by gmaclennan** [comment]
To bring in some ideas from the Mapeo Core API discussion:

- Rename `id` to `docId` and `version` to `versionId`
- Use `updatedAt` instead of `timestamp` (agreed with Seth about snake_case, but let's stick with camel for consistency)

> Store links as Array<{ coreKey: Buffer, index: number }>

I think at some stage we need to encode this to a string, for the client. I think for our sqlite materialized view too. The reason for this would be to save storage/bandwidth. It should be about 32 + 4 = 36 bytes encoded (per link) vs. 64 + 1 + ~4 = 69 bytes as a string. Comes with a CPU cost of encoding to hex on read/write. I don't have a good sense of what the bottle neck is (CPU vs. bandwidth/storage) right now. Maybe easier just to keep as strings? But I'm not sure. I prefer the key as buffer, index as number raw format, because it requires less implicit knowledge (e.g. knowing to split on @, knowing that the first part of the string is hex encoded) if someone is looking at this raw data outside of Mapeo.

> Define type of observation attachments

I think it needs to be:

```ts
{
  id: string,
  type: 'photo' | 'video' | 'audio'
  mimeType: string // IANA mime type
}
```

Maybe the mimeType can be in the blob store metadata? I think it will only be useful for serving blobs over HTTP. To get a blob from the blob store you need: `{ id, type, variant }` - `variant` will depend on context (e.g. screen size)

> Restrictions on tags? E.g. JSON? Nested?

I think theoretically we have supported any JSON for `tags`, but I don't _think_ any nested JSON exists in the wild. You can't create it in Mapeo I don't think. We do have arrays of JSON primitives, as a result of multi-select fields. I think it makes sense to start with:

```ts
type TagArray = Array<TagValue>
type TagValue = boolean | number | string | null | TagArray
```

e.g. a subset of JSON

> timestamps

I think maybe store as number in Protobuf (milliseconds since UNIX epoc), and number for client usage? I feel like it's easier to type-check (as a string you can end up with an invalid string that won't parse as a date - a number can always be parsed as a date). Decoding as a `new Date()` just complicates IPC (since we need to encode/decode across the bridge between processes)

**Comment by gmaclennan** [comment]
Largely fixed by #72. Will create a separate issue for `node`, `way`, `relation`.

**Comment by gmaclennan** [comment]
See ([internal discussion - private link](https://www.notion.so/digidem/Mapeo-Schema-Discussion-51d97f441dae499eaf975f7aa7000829?pvs=4))
